chainerio/filesystems/hdfs.py

```diff
@@ -89,14 +89,25 @@ def info(self):
         info_str = "using hdfs"
         return info_str
 
-    def list(self, path_or_prefix: str = None):
-        if not path_or_prefix or None is path_or_prefix:
+    def list(self, path_or_prefix: str = None, recursive=False):
+        if path_or_prefix is None:
             path_or_prefix = "/user/{}".format(self.username)
 
         self._create_connection()
-        dir_list = self.connection.ls(path_or_prefix)
-        for _dir in dir_list:
-            yield os.path.basename(_dir)
+
+        if recursive:
+            yield from self._recursive_list(path_or_prefix, path_or_prefix)
+        else:
+            dir_list = self.connection.ls(path_or_prefix)
+            for _dir in dir_list:
+                yield os.path.basename(_dir)
+
+    def _recursive_list(self, prefix, path):
+        for _file in self.connection.ls(path):
+            yield _file[_file.find(prefix):]
```

Comment:
What is this code for?
```
>>> s = 'list'
>>> s[s.find('li'):]
'list'
```

Reply:
As explained in ad279e8, this is to convert the full URI to relative path from prefix.
"hdfs://nameservice/prefix_dir/testfile" => "prefix_dir/testfile"